hw/ip/otbn/util/mem_layout.py

```diff
@@ -0,0 +1,135 @@
+'''Simple code that understands enough of otbn.hjson to get a memory layout
+
+Each memory will have an associated "window" entry, which has an offset from
+the start of the IP block's register space. This offset will be treated as an
+LMA for this memory and these memory LMAs will be used uniformly in
+OTBN-specific tooling. This avoids the result depending on the address layout
+of the wider chip.
+
+In particular, note that OTBN ELF binaries will use these LMAs (the VMAs are in
+OTBN's address space, with both IMEM and DMEM starting at 0). A tool that takes
+such binaries and incorporates them into a top-level image will need to do
+address translation (essentially, just adding the base address of the OTBN IP
+block).
+
+'''
+
+import os
+import sys
+from typing import Dict, List, Mapping, Tuple
+
+import hjson  # type: ignore
+
+# We use reggen to read the hjson file. Since that lives somewhere completely
+# different from this script (and there aren't __init__.py files scattered all
+# over the OpenTitan repository), we have to do sys.path hacks to find it.
+_OLD_SYS_PATH = sys.path
+try:
+    _UTIL_PATH = os.path.join(os.path.dirname(__file__),
+                              '..', '..', '..', '..', 'util')
+    sys.path = [_UTIL_PATH] + _OLD_SYS_PATH
+    from reggen.validate import checking_dict, validate   # type: ignore
+finally:
+    sys.path = _OLD_SYS_PATH
```

Comment:
We do have that problem for many of our tools. Can we introduce code to look for the root of the repository (containing the `.git` directory, for example, or add another marker file/directory), and then make everything else relative to that?

Reply:
The problem is that the `sys.path` gymnastics can't go in a library (because you need to do them to find the library!). The rest of the code is pretty boring. If we don't want to encode the relative path of this file in its contents (number of '..' components), you're right: we could search upwards for a `.git` directory or similar. The problem is that *that* code would have to be duplicated in each script too (again: otherwise how do you find the library that tells you how to find the library?).

I'd be very happy to replace this ugliness with something better, but I don't think I know how :-/